<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head>
<title>TopHat :: Center for Bioinformatics and Computational Biology</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<link rel="stylesheet" type="text/css" href="css/style.css" media="screen">
</head>
<body>
<div id="wrap">
  <div id="top">
    <div class="lefts">
    <table cellpadding="2" width="100%">
      <tbody><tr><td>
        <a href="./index.html"><h1>TopHat</h1></a>
        <h2>A spliced read mapper for RNA-Seq</h2>
      </td><td align="right" valign="middle">
        <a href="http://bio.math.berkeley.edu/">
        <img style="vertical-align:middle;padding-top:4px" src="images/UCBerkeley-seal.scaled.gif" border="0">
        </a>&nbsp;
        <a href="http://genomics.jhu.edu/">
        <img style="vertical-align:middle;padding-top:4px" src="images/IGM_logo.png" border="0">
        </a>&nbsp;
        <a href="http://www.cbcb.umd.edu/"><img style="vertical-align:middle;padding-top:4px" src="images/cbcb_logo.gif" border="0"></a>&nbsp;&nbsp;
      </td></tr>
    </tbody></table>
    </div>
  </div>
  
  <div id="main">
	
     <div id="rightside">
 		<h2>Site Map</h2>
		  <div class="box">
		    <ul>
		      <li><a href="index.html">Home</a></li>
		      <li><a href="tutorial.html">Getting started</a></li>
		      <li><a href="manual.html">Manual</a></li>
		      <li><a href="igenomes.html">Index and annotation downloads</a></li>
		      <li><a href="faq.html">FAQ</a></li>
		      <li><a href="http://www.nature.com/nprot/journal/v7/n3/full/nprot.2012.016.html">Protocol</a></li>
		    </ul>
		  </div>
		
		  <h2><u>News and updates</u></h2>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tbody><tr><td>New releases and related tools will be announced through the <a href="https://lists.sourceforge.net/lists/listinfo/bowtie-bio-announce"><b>mailing list</b></a></td></tr>
			  </tbody></table>
		    </ul>
		  </div>
		
		  <h2><u>Getting Help</u></h2>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tbody><tr><td>Questions about TopHat should be sent to <a href="mailto:tophat.cufflinks@gmail.com"><b>tophat.cufflinks@gmail.com</b></a>.  Please do not email technical questions to TopHat contributors directly.</td></tr>
			  </tbody></table>
		    </ul>
		  </div>
		
		  <a href="./downloads"><h2><u>Releases</u></h2></a>
		  <div class="box">
		    <ul>
		      <table width="100%"><tbody><tr>
		      <td>
<br>
</td></tr><tr><td>version 2.0.6 (BETA)</td> <td align="right">11/02/2012</td></tr>
				<tr>
		      		<td><a href="./downloads/tophat-2.0.6.tar.gz" onclick="javascript: pageTracker._trackPageview('/downloads/tophat'); ">&nbsp;&nbsp;&nbsp;Source code</a></td>
		      	</tr>
				<tr>
					<td><a href="./downloads/tophat-2.0.6.Linux_x86_64.tar.gz" onclick="javascript: pageTracker._trackPageview('/downloads/tophat'); ">&nbsp;&nbsp;&nbsp;Linux x86_64 binary</a></td>
				</tr>
				<tr>
					<td><a href="./downloads/tophat-2.0.6.OSX_x86_64.tar.gz" onclick="javascript: pageTracker._trackPageview('/downloads/tophat'); ">&nbsp;&nbsp;&nbsp;Mac OS X x86_64 binary</a></td>
				</tr>
		      
		      
			  </tbody></table>
		    </ul>
		  </div>
		
		  <h2>Related Tools</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://cufflinks.cbcb.umd.edu/">Cufflinks</a>: Isoform assembly and quantitation for RNA-Seq</li>
		      <li><a href="http://bowtie.cbcb.umd.edu">Bowtie</a>: Ultrafast short read alignment</li>
		      <li><a href="http://tophat-fusion.sourceforge.net">TopHat-Fusion</a>: An algorithm for Discovery of Novel Fusion Transcripts</li>
		      <li><a href="http://compbio.mit.edu/cummeRbund/">CummeRbund</a>:
			Visualization of RNA-Seq differential analysis</li>
		    </ul>
		  </div>
		
		  <a href="ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/"><h2><u>Pre-built indexes</u></h2></a>
		  <div class="box">
		    <table width="100%">

		      <!-- hg18 -->

		      <tbody><tr><td>
		        <a href="ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/hg18.ebwt.zip"><i>H. sapiens</i>, UCSC hg18</a>
		        </td><td align="right">
		          <b>2.7 GB</b>
		      </td></tr>
		      <tr><td colspan="2"><table width="100%"><tbody><tr>

		      </tr><tr>
		        <td>&nbsp;colorspace:
		          <a href="ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/hg18_c.ebwt.zip">full</a></td>
		      </tr></tbody></table>
		      </td></tr>

		      <!-- hg19 -->

		      <tr><td>
		        <a href="ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/hg19.ebwt.zip"><i>H. sapiens</i>, UCSC hg19</a>
		        </td><td align="right">
		          <b>2.7 GB</b>
		      </td></tr>
		      <tr><td colspan="2"><table width="100%"><tbody><tr>

		      </tr><tr>
		        <td>&nbsp;colorspace:
		          <a href="ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/hg19_c.ebwt.zip">full</a></td>
		      </tr></tbody></table>

		      <!-- mm9 -->

		      </td></tr><tr><td>
		        <a href="ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/mm9.ebwt.zip"><i>M. musculus</i>, UCSC mm9</a>
		        </td><td align="right">
		          <b>2.4 GB</b>
		      </td></tr>
		      <tr><td colspan="2"><table width="100%"><tbody><tr>
		        

		      </tr><tr>
		        <td>&nbsp;colorspace:
		          <a href="ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/mm9_c.ebwt.zip">full</a></td>
		      </tr></tbody></table>
		      </td></tr>


		  </tbody></table>


	    <p style="font-size: x-small; line-height: 130%">
	    All indexes are for assemblies, not contigs.  Unplaced or unlocalized 
		sequences and alternate haplotype assemblies are excluded.</p>
	    <p style="font-size: x-small; line-height: 130%">
	    Some unzip programs cannot handle archives &gt;2 GB.  If you have
	    problems downloading or unzipping a &gt;2 GB index, try downloading
	    in two parts.</p>
	    <p style="font-size: x-small; line-height: 130%">Check .zip file 
			integrity with <a href="./md5s.shtml">MD5s</a>.</p>

	    <p style="font-size: x-small; line-height: 130%">Pre-built indexes are
		 compatible with Bowtie versions 0.9.8 and later.  For older indexes, please contact us.</p>

	  	  </div>

		  <h2>Publications</h2>
		  <div class="box">
		    <ul>
		      <li style="font-size: x-small; line-height: 130%"><p>Trapnell C, Pachter L, Salzberg SL. <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/btp120"><b>TopHat: discovering splice junctions with RNA-Seq</b></a>. <i><a href="http://bioinformatics.oxfordjournals.org">Bioinformatics</a></i>  doi:10.1093/bioinformatics/btp120</p><br></li>
		      <li style="font-size: x-small; line-height: 130%"><p>Langmead B, Trapnell C, Pop M, Salzberg SL. <a href="http://genomebiology.com/2009/10/3/R25"><b>Ultrafast and memory-efficient alignment of short DNA sequences to the human genome</b></a>. <i><a href="http://genomebiology.com">Genome Biology</a></i> <b>10</b>:R25.</p><br></li>
		      <li style="font-size: x-small; line-height: 130%"><p>Kim D and Salzberg SL. <a href="http://genomebiology.com/2011/12/8/R72/abstract"><b>TopHat-Fusion: an algorithm for discovery of novel fusion transcripts</b></a>. <i><a href="http://genomebiology.com/">Genome Biology</a></i> 2011, 12:R72</p><br></li>
		    </ul>
		  </div>

		  <h2>Contributors</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://www.cs.umd.edu/%7Ecole/">Cole Trapnell</a></li>
			  <li><a href="http://www.umiacs.umd.edu/%7Einfphilo/">Daehwan Kim</a></li>
			  <li>Geo Pertea</li>
			  <li><a href="http://cs.berkeley.edu/%7Epimentel">Harold Pimentel</a></li>
			  <li>Ryan Kelley</li>
		      <li><a href="http://www.math.berkeley.edu/%7Elpachter/">Lior Pachter</a></li>
		      <li><a href="http://www.cbcb.umd.edu/%7Esalzberg/">Steven Salzberg</a></li>
		    </ul>
		  </div>

		  <h2>Links</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://www.umd.edu/">University of Maryland</a></li>
		      <li><a href="http://www.berkeley.edu/">University of California, Berkeley</a></li>
		      <li><a href="http://www.cbcb.umd.edu/">UMD CBCB</a></li>
		      <li><a href="http://www.cs.umd.edu/">UMD Computer Science</a></li>
		      <li><a href="http://www.math.berkeley.edu/">Berkeley Mathematics</a></li>
		      <li><a href="http://www.umiacs.umd.edu/">UMIACS</a></li>
		    </ul>
		  </div>        
          
    </div> <!-- End of "rightside" -->
    <div id="leftside">
  	  <table><tbody><tr><td cellpadding="7">
  	  <h1>Frequently Asked Questions</h1><br>
      <div id="toc">
  	    <ul>
	      
  <li><a href="#edit_dist">How to control the alignment of reads in terms of number of mismatches, gap length etc. ? <br>
</a></li>
  <li><a href="#accuracy">How can I maximize the accuracy of spliced mapping in TopHat?</a></li>

<li><a href="#mate_inner_dist">I don't know the mate inner distance (-r/--mate-inner-dist option) f</a><a href="file:///mylocal/geo/httpd/html/tophat/faq.html#mate_inner_dist">or my paired reads</a><a href="#mate_inner_dist">, what value should I use?</a></li>

 	      <li><a href="#library_type">I am not sure which library type to use (fr-firststrand or fr-secondstrand), what should I do?</a></li>
  	    </ul><br>
  <h2 id="edit_dist">How to control the alignment of reads in terms of number of mismatches, gap length etc. ? <br>
  </h2>
  <p><br>
You can use three options:<tt> --read-mismatches</tt>, <tt>--read-gap-length</tt> and  
  <tt>--read-edit-dist</tt>. For instance, if you want read alignments with at most 2 base
mismatches and no gaps then you can specify:<br>
&nbsp; <tt>--read-mismatches 2 
--read-gap-length 0 --read-edit-dist 2</tt><br>
Or if you 
want read alignments with total length of indels (alignment gaps) of at 
most 3bp and at most 2 base mismatches you can use these options:<br>
  <tt>&nbsp;--read-mismatches 2 
--read-gap-length 3 --read-edit-dist 3</tt><br>
  </p>
  <p><br>
  </p>
  
  <h2 id="accuracy">How can I maximize the accuracy of spliced mapping in TopHat?<br>
  </h2>

<br>
	Based on real RNA-seq samples we found out that in the genome mapping 
step of TopHat a high portion of reads spanning several exons can 
incorrectly be aligned to processed pseudogenes that are rarely (if any)
 transcribed or expressed, instead of the genes where they originate 
from. You can use either of the options below to improve the accuracy of spliced mapping in TopHat:<br>
<ul>
  <li>If a good gene annotation is available (as the
 case with the human genome), use it with the -G option<br>
  </li>
  
  <li>For poorly annotated genomes you might want to consider using the "<tt>--read-realign-edit-dist 0</tt>" option <br>
With the realignment option
users can choose to remap some (or all) of the mapped reads with mapping edit distance
equal to or above user-specified "remapping" edit distance (see <a href="manual.html#read_realign">--read-realign-edit-dist</a> option). Setting "<tt>--read-realign-edit-dist 0</tt>"
will map every read against transcriptome, genome, and splice variants (or splice junctions)
that are detected by TopHat, no matter whether it is mapped or not in any mapping step. With this
remapping strategy, this "pseudogene" problem can be effectively
handled. If you use a genome that has processed pseudogenes
and you cannot provide good gene annotation to TopHat, you may want to consider using
this option for accurate mapping results.
	<ul>
<ul>

</ul>&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;  <br>
<br>
  	  </ul></li>
</ul></div>
      
  <h2 id="mate_inner_dist">I don't know the mate inner distance (-r/--mate-inner-dist) for my paired reads, what value should I use?</h2>
<br><p>The 
default value should work fine in most cases, for typical RNA-Seq PE experiments, because TopHat allows some variance for this distance internally.&nbsp; 
TopHat makes use of the mate inner distance information in several 
places - for instance, when finding splice sites and fusion break 
points. This information is also taken into account when choosing the 
best candidate alignments for paired reads in the final stage of TopHat 
(tophat_reports). If you want to find a good approximation of this 
distance for your reads you can try running Bowtie2 on a small sample 
(subset) of the paired reads (both mates) 
and&nbsp; taking a look at their mapped positions (we hope to add this 
automatic fragment length detection in a future version of TopHat). The SAM output of Bowtie2 for paired reads is especially helpful as the 9th field in the SAM alignment lines should show the estimated fragment length, from which you should subtract twice the read length to get the value of the "inner distance" that can be used with the -r parameter (obviously large absolute values for that field should be ignored as for this estimate we only want to consider mates aligned to the same exon).<br>
</p><br>
      <h2 id="library_type">I am not sure which library type to use (fr-firststrand or fr-secondstrand), what should I do?</h2><br><p>
	    One possible way to figure out the correct library-type is to run 
TopHat with a small subset of the reads (e.g., 1M) as follows.&nbsp;</p>
<ol>
  <li>run TopHat with fr-firststrand and count the number of junctions in junctions.bed (one of the output files from TopHat)</li>
  <li>run TopHat with fr-secondstrand and count the number of junctions in junctions.bed&nbsp;</li>
</ol>
<p>
Since the splice junction finding algorithm of TopHat makes use of 
library-type information (if provided), one of the two TopHat runs would
 result in many more splice junctions than the other one.
You can then use the library type that gives more junctions.
If this is not the case TopHat might not work well with your sequencing 
protocol.  Please let us know more details about your protocol so we can
 add support for new library types. 
  	  </p>
<br><br>

	
  
  <div id="footer">
  	<table cellspacing="15" width="100%"><tbody><tr><td>
    This research was supported in part by NIH grants R01-LM06845 and R01-GM083873 and NSF grant CCF-0347992.
    </td><td align="right">
    Administrator: <a href="mailto:cole@cs.umd.edu">Cole Trapnell</a>. Design by <a href="http://www.free-css-templates.com" title="Design by David Herreman">David Herreman</a><br>
<br>

    </td></tr></tbody></table>
  </div>




<!-- Google analytics code -->
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
var pageTracker = _gat._getTracker("UA-6101038-1");
pageTracker._trackPageview();
</script>




</td></tr></tbody></table></div></div></div>








</body></html>
