<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>TopHat-Fusion :: Center for Bioinformatics and Computational Biology</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<link rel="stylesheet" type="text/css" href="css/style.css" media="screen" />
</head>
<body>
<div id="wrap">
  <div id="top">
    <div class="lefts">
    <table width="100%" cellpadding="2">
      <tr><td>
           <a href="./fusion_index.html"><h1>TopHat-Fusion</h1></a>
              <h2>An algorithm for Discovery of Novel Fusion Transcripts</h2>
      </td><td align="right" valign="middle">
        <a href="http://www.mcb.berkeley.edu/"><img style="vertical-align:middle;padding-top:4px" border=0 src="images/UCBerkeley-seal.scaled.gif"></img></a>&nbsp;
        <a href="http://www.cbcb.umd.edu/"><img style="vertical-align:middle;padding-top:4px" border=0 src="images/cbcb_logo.gif"></a>&nbsp;&nbsp;
      </td></tr>
      </td></tr>
    </table>
    </div>
  </div>

  <div id="main">
	
    <div id="rightside">
		<h2>Site Map</h2>
		  <div class="box">
		    <ul>
		      <li><a href="fusion_index.html">Home</a></li>
		      <li><a href="fusion_tutorial.html">Getting started</a></li>
		      <li><a href="fusion_manual.html">Manual</a></li>
		    </ul>
		  </div>
		
    </div> <!-- End of "rightside" -->
    <div id="leftside">
  	<table><tr><td cellpadding=7>
  	<h1> Getting started</h1><br/>
      <div id="toc">
  	    <ul>
	      <li><a href="#anno">Downloading annotations</a></li>
  	      <li><a href="#toph">Running TopHat-Fusion</a></li>
  	      <li><a href="#exam">Examining your results</a></li>
  	    </ul><br/>
  	  </div>

      	<br/>
      	<h2 id="anno">Downloading annotations</h2><br/>
	<p> 
	</p>
	    <TABLE id="box-table-a" CELLPADDING=3>
      <TR>
<TH ALIGN="LEFT">Organism</TH>
<TH ALIGN="LEFT">Data source</TH>
<TH ALIGN="LEFT">Version</TH>
<TH ALIGN="LEFT">Bowtie indexes</TH>
<TH ALIGN="LEFT">Annotation</TH>
</TR>
<TR>
<TD ALIGN="LEFT" rowspan=2>Homo sapiens</TD>
<TD ALIGN="LEFT" rowspan=2>UCSC</TD>
<TD ALIGN="LEFT">hg18</TD>
<TD ALIGN="RIGHT">
  <a href = "ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/hg18.ebwt.zip">Bowtie1</a> /
  <a href = "ftp://ftp.cbcb.umd.edu/pub/data/bowtie2_indexes/hg18.zip">Bowtie2</a>
</TD>
<TD ALIGN="LEFT">
  <a href = "http://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/refGene.txt.gz">refGene.txt</a> /
  <a href = "http://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/ensGene.txt.gz">ensGene.txt</a>
</TD>
</TR>
<TR>
<TD ALIGN="LEFT">hg19</TD>
<TD ALIGN="RIGHT">
  <a href = "ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/hg19.ebwt.zip">Bowtie1</a> /
  <a href = "ftp://ftp.cbcb.umd.edu/pub/data/bowtie2_indexes/hg19.zip">Bowtie2</a>
</TD>
<TD ALIGN="LEFT">
  <a href = "http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/refGene.txt.gz">refGene.txt</a> /
  <a href = "http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/ensGene.txt.gz">ensGene.txt</a>
</TD>
</TR>
<TR>
<TD ALIGN="LEFT" rowspan=1>Mus musculus</TD>
<TD ALIGN="LEFT" rowspan=1>UCSC</TD>
<TD ALIGN="LEFT">mm9</TD>
<TD ALIGN="RIGHT">
  <a href = "ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/mm9.ebwt.zip">Bowtie1</a> /
  <a href = "ftp://ftp.cbcb.umd.edu/pub/data/bowtie2_indexes/mm9.zip">Bowtie2</a>
</TD>
<TD ALIGN="LEFT">
  <a href = "http://hgdownload.cse.ucsc.edu/goldenPath/mm9/database/refGene.txt.gz">refGene.txt</a> /
  <a href = "http://hgdownload.cse.ucsc.edu/goldenPath/mm9/database/ensGene.txt.gz">ensGene.txt</a>
</TD>
</TR>
<TR>
<TD ALIGN="LEFT" rowspan=1>Rattus norvegicus</TD>
<TD ALIGN="LEFT" rowspan=1>UCSC</TD>
<TD ALIGN="LEFT">rn4</TD>
<TD ALIGN="RIGHT">
  <a href = "ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/rn4.ebwt.zip">Bowtie1</a>
</TD>
<TD ALIGN="LEFT">
  <a href = "http://hgdownload.cse.ucsc.edu/goldenPath/rn4/database/refGene.txt.gz">refGene.txt</a> /
  <a href = "http://hgdownload.cse.ucsc.edu/goldenPath/rn4/database/ensGene.txt.gz">ensGene.txt</a>
</TD>
</TR>
<!--
<TR>
<TD ALIGN="LEFT" rowspan=1>Bos taurus</TD>
<TD ALIGN="LEFT" rowspan=1>UCSC</TD>
<TD ALIGN="LEFT">bosTau4</TD>
<TD ALIGN="RIGHT">
  <a href = "">Bowtie1</a>
</TD>
<TD ALIGN="LEFT">Oct 20 2011</TD>
</TR>
-->
<TR>
<TD ALIGN="LEFT" rowspan=1>Canis familiaris</TD>
<TD ALIGN="LEFT" rowspan=1>UCSC</TD>
<TD ALIGN="LEFT">canFam2</TD>
<TD ALIGN="RIGHT">
  <a href = "ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/canFam2.ebwt.zip">Bowtie1</a>
</TD>
<TD ALIGN="LEFT">
  <a href = "http://hgdownload.cse.ucsc.edu/goldenPath/canFam2/database/refGene.txt.gz">refGene.txt</a> /
  <a href = "http://hgdownload.cse.ucsc.edu/goldenPath/canFam2/database/ensGene.txt.gz">ensGene.txt</a>
</TD>
</TR>
<TR>
<TD ALIGN="LEFT" rowspan=1>Gallus gallus</TD>
<TD ALIGN="LEFT" rowspan=1>UCSC</TD>
<TD ALIGN="LEFT">galGal3</TD>
<TD ALIGN="RIGHT">
  <a href = "ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/galGal3.ebwt.zip">Bowtie1</a>
</TD>
<TD ALIGN="LEFT">
  <a href = "http://hgdownload.cse.ucsc.edu/goldenPath/galGal3/database/refGene.txt.gz">refGene.txt</a> /
  <a href = "http://hgdownload.cse.ucsc.edu/goldenPath/galGal3/database/ensGene.txt.gz">ensGene.txt</a>
</TD>
</TR>
</TABLE>

	<br/>

	  	
	<br/>
      	<h2 id="toph">Running TopHat-Fusion</h2><br/>
	<p> TopHat-Fusion is built on <a href="http://tophat.cbcb.umd.edu">TopHat</a>
	  so that it inherits every option and output formats from TopHat
	  (Refer to the TopHat website for installation and basic information).
	  TopHat-Fusion algorithm  is described in <a href="./data/TopHatFusion-poster.pdf">our poster</a> at the CSHL Biology of Genomes conference.
	  TopHat-Fusion consists of two sub-programs (<tt><b>tophat</b></tt> and <tt><b>tophat-fusion-post</b></tt>).
	  Using a breast cancer cell MCF7 RNA-Seq data from Edgren et al (Genome Biology 2011). </a>,
	  the following tutorial demonstrates how to use TopHat-Fusion to identify fusion genes
	  including three known fusions (BCAS4-BCAS3, ARFGEF2-SULF2, RPS6KB1-TMEM49).
	</p>
	<br/>

	<p> MCF7 RNA-Seq data (paired-end 50-bp) can be downloaded
	  either from <a href="ftp://ftp.cbcb.umd.edu/pub/data/tophat-fusion/SRR064286.tar.gz">our ftp server</a>
	  or from <a href="http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?cmd=viewer&m=data&s=viewer&run=SRR064286">NCBI Sequence Read Archive</a>.</p>
        <p> To run TopHat-Fusion: </p>
        
        <br/>
        <blockquote>
	  <b>
	    tophat -o tophat_MCF7 -p 8 --fusion-search --keep-fasta-order --bowtie1
	    --no-coverage-search -r 0 --mate-std-dev 80
	    --fusion-min-dist 100000 --fusion-anchor-length 13 --fusion-ignore-chromosomes chrM
	    /path/to/h_sapiens/bowtie_index SRR064286_1.fastq SRR064286_2.fastq
	  </b>
	    
	  <br><br>
	  <ul>
  	    <li>
	      Make (top_dir) directory and run the above command under (top_dir) - see <a href="#directory">suggested directory structure.</a> If you have multiple samples, you can run them under (top_dir).
	    </li>
  	    <li>
	      Use tophat_(sample_name) for the output directory ("-o" option) such as tophat_MCF7.
	      The directory name (MCF7) will be used later for annotation purposes.
	    </li>
	    <li> You can change the number of threads using "-p" option. </li>
	    <li> Turn on fusion algorithm (--fusion-search) and use Bowtie1 (--bowtie1). </li>
	    <li> Turn off "coverage-search", which takes lots of memory and is slow. </li>
	    <li>
	      The mean fragment length of the data is 100-bp, so the inner mate distance is 0 (= 100 - 50 * 2).
	      We use a somewhat big mate standard derivation (80-bp)
	      because TopHat-Fusion makes use of the area (mate_inner_dist &plusmn; std_dev) when it tries to find fusions.
	    </li>
	    <li>
	      In addition to inter-chromosomal fusions,
	      TopHat-Fusion tries to identify intra-chromosomal fusions
	      due to rearrangement within a chromosome separated by at least --fusion-min-dist.
	    </li>
	    <li>
	      A read supports a fusion if a read maps to both sides of a fusion by at least --fusion-anchor-length.
	    </li>
	    <li>
	      In addition to <a href="http://tophat.cbcb.umd.edu/manual.html#output">outputs</a> from TopHat,
	      TopHat-Fusion outputs a list of potential fusions (<a href="./data/fusions.out">fusions.out</a> - the first 2,000 out of 68,168 fusions) and
	      a modified SAM alignment that allows "fusion" alignment using 'F' CIGAR operator
	      although it is not supported by SAM tools.
	    </li>
  	  </ul>
	</blockquote>
        <br/>

        <br/>
        <blockquote>
	  <b>
	    tophat-fusion-post -p 8 --num-fusion-reads 1 --num-fusion-pairs 2 --num-fusion-both 5 /path/to/h_sapiens/bowtie_index
	  </b>

	  <br><br>
	  <ul>
  	    <li> Copy (tophat_install_dir)/annotation/* into (top_dir). </li>
	    <li>
	      TopHat-Fusion uses BLAST search results for filtering out false fusions due to highly similar sequences or pseudogenes.
	      Also, the search results can be alternatively used for annotating purposes in case there is no known genes in the provided annotation files.
	      50-bp sequence on the left side of a fusion and 50-bp on the right side are combined to make a 100-bp sequence, which in turn is BLASTed against human genome.
	      If match length (range: 0 to 100) + identity percent (0 to 100) is greater than 160, the fusion is filtered out.
	      This BLAST step is usually done for a few hundreds of fusions after prior filtering steps.
	      Thus, it is highly recommended to install BLAST and download blast database as follows.
	      <ul>
		<li> Install <a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download">BLAST binaries</a> (blastall and blastn). </li>
		<li>
		  Make (top_dir)/blast directory,
		  download human_genomic*, other_genomic*, and nt* from <a href="ftp://ftp.ncbi.nlm.nih.gov/blast/db/">blast database</a>, and
		  extract them under (top_dir)/blast_human.
		</li>
	      </ul>
	    </li>
	    <li>
	      The final list of fusion candidates is given in <a href="./data/result.html">(top_dir)/tophatfusion_out/result.html</a>.
	    </li>
	    <li>
	      You may want to repeat the filtering process with various filtering parameters such as --num-fusion-reads and --num-fusion-pairs
	      without deleting (top_dir)/tophatfusion_out, which is a database <tt><b>tophat-fusion-post</b><tt> internally uses for fast computation.
	    </li>
  	  </ul>
	</blockquote>
        <br/>

	<blockquote>
	  <b id="directory">Suggested directory structure</b> <br/> <br/>
	  <ul>
	    <li> (top_dir)
	      <ul>
		<li> tophat_sample_1 - the output directory by <tt><b>tophat</b> and you may want to run it on several samples.</tt> </li>
		<li> tophat_sample_2 </li>
		<li> ... </li>
		<li> tophat_sample_n </li> <br/>
		<li> tophatfusion_out - the output directory by <tt><b>tophat-fusion-post</b></tt> </li> <br/>
		<li> ensGene.txt </li>
		<li> refGene.txt </li>
		<li> <a href = "data/mcl">mcl</a> </li>
		<li> blast - BLAST database </li>
	      </ul>
	    </li>
	  </ul>
	</blockquote>
	<br/><br/>

      	<h2 id="exam">Examining your output</h2><br/>
      	<p>TopHat-Fusion produces <a href="./fusion_manual.html#output">several files</a>.
	</p>
	<br/>

  <div id="footer">
  	<table width="100%" cellspacing=15><tr><td>
    This research was supported in part by NIH grants R01-LM06845 and R01-GM083873.
    </td><td align=right>
    Administrator: <a href="mailto:infphilo@umiacs.umd.edu">Daehwan Kim</a>. Design by <a href="http://www.free-css-templates.com" title="Design by David Herreman">David Herreman</a>
    </td></tr></table>
  </div>
</div>

<!-- Google analytics code -->
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
var pageTracker = _gat._getTracker("UA-6101038-1");
pageTracker._trackPageview();
</script>

<!-- End google analytics code -->

</body>
</html>
