<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>TopHat-Fusion :: Center for Bioinformatics and Computational Biology</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<link rel="stylesheet" type="text/css" href="css/style.css" media="screen" />
</head>
<body>
<div id="wrap">
  <div id="top">
   <div class="lefts">
    <table width="100%" cellpadding="2">
      <tr><td>
        <a href="./fusion_index.html"><h1>TopHat-Fusion</h1></a>
        <h2>An algorithm for Discovery of Novel Fusion Transcripts</h2>
      </td><td align="right" valign="middle">
        <a href="http://www.mcb.berkeley.edu/"><img style="vertical-align:middle;padding-top:4px" border=0 src="images/UCBerkeley-seal.scaled.gif"></img></a>&nbsp;
	<a href="http://genomics.jhu.edu/">
        <img style="vertical-align:middle;padding-top:4px"
            src="images/IGM_logo.png" border="0">
        <a href="http://www.cbcb.umd.edu/"><img style="vertical-align:middle;padding-top:4px" border=0 src="images/cbcb_logo.gif"></a>&nbsp;&nbsp;
      </td></tr>
    </table>
    </div>
  </div>

  <div id="subheader">
  <table width="100%"><tr>
  <td>
  <strong>TopHat-Fusion</strong> is an enhanced version of TopHat with the ability to align reads across fusion points, which results from the breakage and re-joining of two different chromosomes, or from rearrangements within a chromosome. 
  	<br/> <br/>
  </td><td align=right valign=middle>
  <a href="http://opensource.org"><img border=0 alt="Open Source Software" src="images/osi-certified.gif"></img></a>
  </td></tr>
  </table>
  </div>
  <div id="main">
    <div id="rightside">
	
		<h2>Site Map</h2>
		  <div class="box">
		    <ul>
		      <li><a href="fusion_index.html">Home</a></li>
		      <li><a href="fusion_tutorial.html">Getting started</a></li>
		      <li><a href="fusion_manual.html">Manual</a></li>
		    </ul>
		  </div>
		
		  <h2><u>News and updates</u></h2>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tr><td>New releases and related tools will be announced through the <a href="https://lists.sourceforge.net/lists/listinfo/bowtie-bio-announce"><b>mailing list</b></a></td></tr>
			  </table>
		    </ul>
		  </div>
		
		  <h2><u>Getting Help</u></h2>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tr><td>Questions about TopHat-Fusion should be sent to <a href="mailto:tophat.cufflinks@gmail.com"><b>tophat.cufflinks@gmail.com</b></a>.
				    Please do not email technical questions to TopHat-Fusion contributors directly.</td></tr>
			  </table>
		    </ul>
		  </div>
		
		  <h2>Related Tools</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://cufflinks.cbcb.umd.edu/">Cufflinks</a>: Isoform assembly and quantitation for RNA-Seq</li>
		      <li><a href="http://tophat.cbcb.umd.edu">TopHat</a>: A spliced read mapper for RNA-Seq</li>
		      <li><a href="http://bowtie.cbcb.umd.edu">Bowtie</a>: Ultrafast short read alignment</li>
		    </ul>
		  </div>
		
		  <h2>Publications</h2>
		  <div class="box">
		    <ul>
		      <li style="font-size: x-small; line-height: 130%"><p>Trapnell C, Pachter L, Salzberg SL. <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/btp120"><b>TopHat: discovering splice junctions with RNA-Seq</b></a>. <i><a href="http://bioinformatics.oxfordjournals.org">Bioinformatics</a></i>  doi:10.1093/bioinformatics/btp120</p><br/></li>
		      <li style="font-size: x-small; line-height: 130%"><p>Kim D and Salzberg SL. <a href="http://genomebiology.com/2011/12/8/R72/abstract"><b>TopHat-Fusion: an algorithm for discovery of novel fusion transcripts</b></a>. <i><a href="http://genomebiology.com/">Genome Biology</a></i> 2011, 12:R72</p><br/></li>
		    </ul>
		  </div>

		  <h2>Contributors</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://www.umiacs.umd.edu/~infphilo/">Daehwan Kim</a></li>
		      <li><a href="http://www.cs.umd.edu/~cole/">Cole Trapnell</a></li>
		      <li>Geo Pertea</li>
		      <li>Ryan Kelley</li>
		      <li><a href="http://www.math.berkeley.edu/~lpachter/">Lior Pachter</a></li>
		      <li><a href="http://www.cbcb.umd.edu/~salzberg/">Steven Salzberg</a></li>
		    </ul>
		  </div>

		  <h2>Links</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://www.umd.edu/">University of Maryland</a></li>
		      <li><a href="http://www.berkeley.edu/">University of California, Berkeley</a></li>
		      <li><a href="http://www.cbcb.umd.edu/">UMD CBCB</a></li>
		      <li><a href="http://www.cs.umd.edu/">UMD Computer Science</a></li>
		      <li><a href="http://www.math.berkeley.edu/">Berkeley Mathematics</a></li>
		      <li><a href="http://www.umiacs.umd.edu/">UMIACS</a></li>
		    </ul>
		  </div>
		
    </div> <!-- End of "rightside" -->
    <div id="leftside">
      <h2> TopHat2 is released in which TopHat-Fusion is included (04/06/2012) </h2>
      <p>
	<ul>
	  <li> Now TopHat-Fusion engine is incorported into TopHat2 with the name of --fusion-search option,
	    benefiting from TopHat2's advanced features. </li>
	  <li> Many bug fixes include some incompatibility issues with Python and memory problem in tophat_reports. </li>
	  <li> <a href = "fusion_tutorial.html#anno">Annotations</a> for Mouse, Rat, Dog, and Chicken are added.
	    Please let us know if you need annotations for other genomes.</li>
	  <li> Fusion alignment format is slightly changed in a way compatible with SAM using optional fields "XF" and "XP"
	    (see <a href = "fusion_manual.html#output">the output format</a> for more detail) </li>
	  <li> In this version, you may want to use the older version of Bowtie (--bowtie1),
	    which is much faster than with Bowtie2.</li>
	  </ul>
      </p>
      <br/>

      <h2> Important notes (12/19/2011) </h2>
      <p>
	<ul>
	  <li>
	    If you have both TopHat and TopHat-Fusion installed on the same machine,
	    you may need to put the path for TopHat-Fusion in your shell's PATH so that the sub-programs included in the TopHat-Fusion's package are used.
	  </li>
	  <li>
	    We have been working on the second version of TopHat-Fusion, which (we hope) will be available <s>before March, 2012</s> in April, 2012.
	  </li>
	  </ul>
      </p>
      <br/>

      <h2> Important notes (10/07/2011) </h2>
      <p>
	<ul>
	  <li>
	    When installing TopHat-Fusion, if you already had TopHat (upon which TopHat-Fusion is based) installed somewhere,
	    please make sure you install (or copy) all programs in TopHat-Fusion (tophat-fusion-0.1.0/bin) in a directory different than the one the binaries of TopHat are installed in.
	    Otherwise, TopHat-Fusion will replace the binaries of TopHat with its own ones (tophat.py, segment_juncs, or etc).
	  </li>
	  <li>
	    The current version of TopHat-Fusion has been tested using Python 2.7.1, so if you are using a different version of Python (2.6.X or 3.X), you are encouraged to try Python 2.7.1.
	  </li>
	  </ul>
      </p>
      <br/>
      <h2> TopHat-Fusion paper published (8/11/2011) </h2>
      <p>
	The TopHat-Fusion <a href="http://genomebiology.com/2011/12/8/R72/abstract">paper</a> has been published in <a href="http://genomebiology.com/"><i>Genome Biology</i></a>.
      </p>
      <br/>
      <h2> TopHat-Fusion 0.1.0 (Beta) is released (5/09/2011) </h2>
      <p>
	The TopHat-Fusion algorithms are described in <a href="./data/TopHatFusion-poster.pdf">our poster</a> at the CSHL Biology of Genomes conference.
	<ul>
	  <li>
	    We suggest that reads are at least 50-bp long, where a read is split into two segments (25-bp each)
	    and at least two segments are required for identifying fusions.
	    Shorter reads can be used if segment-length is set to a smaller value.
	  </li>
	  <li>
	    Both single and paired-end reads can be used and the output alignments are given in a modified SAM format
	    where we added a new CIGAR<sup>*</sup> operator 'F' to indicate fusion points.
	    (See <a href="./fusion_manual.html#output">Output</a> for detailed information.)
	    <h5><blockquote>
	      * CIGAR is an acronym for Concise Idiosyncratic Gapped Alignment Report.
	    </blockquote></h5>
	  </li>
	  <li> This version does not support color-space reads although we have a plan to support it soon. </li>
	  <li>
	    TopHat-Fusion consists of two sub-programs: tophat-fusion and tophat-fusion-post. For more information,
	    please refer to <a href="./fusion_tutorial.html">Getting started</a>.
	  </li>
	  <li>
	    This version is focused on identifying fusions due to chromosomal rearragements whether they are inter- or intra-chromosomal,
	    but fusions due to read-through transcription or trans-splicing will be eventually supported.
	    In addition to fusion read alignments, discordantly mapped pairs are output as well.
	  </li>
	  <li>
	    As an illustration, a sample output of TopHat-Fusion is given in <a href="./data/result.html">HTML format</a>.
	    See <a href="./fusion_tutorial.html">Getting started</a>.
	  </li>
	</ul>
      </p>
      <br/>
      </div>
    </div>
    <div id="footer">
  	<table width="100%" cellspacing=15><tr><td>
    This research was supported in part by NIH grants R01-LM06845 and R01-GM083873.
    </td><td align=right>
    Administrator: <a href="mailto:infphilo@umiacs.umd.edu">Daehwan Kim</a>. Design by <a href="http://www.free-css-templates.com" title="Design by David Herreman">David Herreman</a>
    </td></tr></table>
  </div>
</div>

<!-- Google analytics code -->
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
var pageTracker = _gat._getTracker("UA-6101038-1");
pageTracker._trackPageview();
</script>
<!-- End google analytics code -->

</body>
</html>
