<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head>
<title>TopHat :: Center for Bioinformatics and Computational Biology</title>


<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">


<link rel="stylesheet" type="text/css" href="css/style.css" media="screen">


</head>
<body>
<div id="wrap">
  <div id="top">
    <div class="lefts">
    <table cellpadding="2" width="100%">
      <tbody><tr><td>
        <a href="./index.html"><h1>TopHat</h1></a>
        <h2>A spliced read mapper for RNA-Seq</h2>
      </td><td align="right" valign="middle">
        <a href="http://bio.math.berkeley.edu/">
        <img style="vertical-align:middle;padding-top:4px" src="images/UCBerkeley-seal.scaled.gif" border="0">
        </a>&nbsp;
        <a href="http://genomics.jhu.edu/">
        <img style="vertical-align:middle;padding-top:4px" src="images/JHU-seal.gif" border="0">
        </a>&nbsp;
        <a href="http://www.cbcb.umd.edu/"><img style="vertical-align:middle;padding-top:4px" src="images/cbcb_logo.gif" border="0"></a>&nbsp;&nbsp;
      </td></tr>
    </tbody></table>
    </div>
  </div>
  
  <div id="main">
	
     <div id="rightside">
 		<h2>Site Map</h2>
		  <div class="box">
		    <ul>
		      <li><a href="index.html">Home</a></li>
		      <li><a href="tutorial.html">Getting started</a></li>
		      <li><a href="manual.html">Manual</a></li>
		      <li><a href="igenomes.html">Index and annotation downloads</a></li>
		      <li><a href="http://www.nature.com/nprot/journal/v7/n3/full/nprot.2012.016.html">Protocol</a></li>
		    </ul>
		  </div>
		
		  <h2><u>News and updates</u></h2>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tbody><tr><td>New releases and related tools will be announced through the <a href="https://lists.sourceforge.net/lists/listinfo/bowtie-bio-announce"><b>mailing list</b></a></td></tr>
			  </tbody></table>
		    </ul>
		  </div>
		
		  <h2><u>Getting Help</u></h2>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tbody><tr><td>Questions about TopHat should be sent to <a href="mailto:tophat.cufflinks@gmail.com"><b>tophat.cufflinks@gmail.com</b></a>.  Please do not email technical questions to TopHat contributors directly.</td></tr>
			  </tbody></table>
		    </ul>
		  </div>
		
		  <a href="./downloads"><h2><u>Releases</u></h2></a>
		  <div class="box">
		    <ul>
		      <table width="100%"><tbody><tr>
		      <td>
				<br>
</td></tr><tr><td>version 2.0.0 (BETA)</td> <td align="right">2/23/2012</td></tr>
				<tr>
		      		<td><a href="./downloads/tophat-2.0.0.tar.gz" onclick="javascript: pageTracker._trackPageview('/downloads/tophat'); ">&nbsp;&nbsp;&nbsp;Source code</a></td>
		      	</tr>
				<tr>
					<td><a href="./downloads/tophat-2.0.0.Linux_x86_64.tar.gz" onclick="javascript: pageTracker._trackPageview('/downloads/tophat'); ">&nbsp;&nbsp;&nbsp;Linux x86_64 binary</a></td>
				</tr>
				<tr>
					<td><a href="./downloads/tophat-2.0.0.OSX_x86_64.tar.gz" onclick="javascript: pageTracker._trackPageview('/downloads/tophat'); ">&nbsp;&nbsp;&nbsp;Mac OS X x86_64 binary</a></td>
				</tr>
		      
		      
			  </tbody></table>
		    </ul>
		  </div>
		
		  <h2>Related Tools</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://cufflinks.cbcb.umd.edu/">Cufflinks</a>: Isoform assembly and quantitation for RNA-Seq</li>
		      <li><a href="http://bowtie.cbcb.umd.edu">Bowtie</a>: Ultrafast short read alignment</li>
		      <li><a href="http://tophat-fusion.sourceforge.net">TopHat-Fusion</a>: An algorithm for Discovery of Novel Fusion Transcripts</li>
		      <li><a href="http://compbio.mit.edu/cummeRbund/">CummeRbund</a>:
			Visualization of RNA-Seq differential analysis</li>
		    </ul>
		  </div>
		
		  <a href="ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/"><h2><u>Pre-built indexes</u></h2></a>
		  <div class="box">
		    <table width="100%">

		      <!-- hg18 -->

		      <tbody><tr><td>
		        <a href="ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/hg18.ebwt.zip"><i>H. sapiens</i>, UCSC hg18</a>
		        </td><td align="right">
		          <b>2.7 GB</b>
		      </td></tr>
		      <tr><td colspan="2"><table width="100%"><tbody><tr>

		      </tr><tr>
		        <td>&nbsp;colorspace:
		          <a href="ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/hg18_c.ebwt.zip">full</a></td>
		      </tr></tbody></table>
		      </td></tr>

		      <!-- hg19 -->

		      <tr><td>
		        <a href="ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/hg19.ebwt.zip"><i>H. sapiens</i>, UCSC hg19</a>
		        </td><td align="right">
		          <b>2.7 GB</b>
		      </td></tr>
		      <tr><td colspan="2"><table width="100%"><tbody><tr>

		      </tr><tr>
		        <td>&nbsp;colorspace:
		          <a href="ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/hg19_c.ebwt.zip">full</a></td>
		      </tr></tbody></table>

		      <!-- mm9 -->

		      </td></tr><tr><td>
		        <a href="ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/mm9.ebwt.zip"><i>M. musculus</i>, UCSC mm9</a>
		        </td><td align="right">
		          <b>2.4 GB</b>
		      </td></tr>
		      <tr><td colspan="2"><table width="100%"><tbody><tr>
		        

		      </tr><tr>
		        <td>&nbsp;colorspace:
		          <a href="ftp://ftp.cbcb.umd.edu/pub/data/bowtie_indexes/mm9_c.ebwt.zip">full</a></td>
		      </tr></tbody></table>
		      </td></tr>


		  </tbody></table>


	    <p style="font-size: x-small; line-height: 130%">
	    All indexes are for assemblies, not contigs.  Unplaced or unlocalized 
		sequences and alternate haplotype assemblies are excluded.</p>
	    <p style="font-size: x-small; line-height: 130%">
	    Some unzip programs cannot handle archives &gt;2 GB.  If you have
	    problems downloading or unzipping a &gt;2 GB index, try downloading
	    in two parts.</p>
	    <p style="font-size: x-small; line-height: 130%">Check .zip file 
			integrity with <a href="./md5s.shtml">MD5s</a>.</p>

	    <p style="font-size: x-small; line-height: 130%">Pre-built indexes are
		 compatible with Bowtie versions 0.9.8 and later.  For older indexes, please contact us.</p>

	  	  </div>

		  <h2>Publications</h2>
		  <div class="box">
		    <ul>
			  <li style="font-size: x-small; line-height: 130%"><p>Trapnell C, Pachter L, Salzberg SL. <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/btp120"><b>TopHat: discovering splice junctions with RNA-Seq</b></a>. <i><a href="http://bioinformatics.oxfordjournals.org">Bioinformatics</a></i>  doi:10.1093/bioinformatics/btp120</p><br></li>
		      <li style="font-size: x-small; line-height: 130%"><p>Langmead B, Trapnell C, Pop M, Salzberg SL. <a href="http://genomebiology.com/2009/10/3/R25"><b>Ultrafast and memory-efficient alignment of short DNA sequences to the human genome</b></a>. <i><a href="http://genomebiology.com">Genome Biology</a></i> <b>10</b>:R25.</p></li>
		    </ul>
		  </div>

		  <h2>Contributors</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://www.cs.umd.edu/%7Ecole/">Cole Trapnell</a></li>
			  <li><a href="http://www.cs.umd.edu/%7Einfphilo/">Daehwan Kim</a></li>
			  <li>Geo Pertea</li>
			  <li><a href="http://cs.berkeley.edu/%7Epimentel">Harold Pimentel</a></li>
			  <li>Ryan Kelley</li>
		      <li><a href="http://www.math.berkeley.edu/%7Elpachter/">Lior Pachter</a></li>
		      <li><a href="http://www.cbcb.umd.edu/%7Esalzberg/">Steven Salzberg</a></li>
		    </ul>
		  </div>

		  <h2>Links</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://www.umd.edu/">University of Maryland</a></li>
		      <li><a href="http://www.berkeley.edu/">University of California, Berkeley</a></li>
		      <li><a href="http://www.cbcb.umd.edu/">UMD CBCB</a></li>
		      <li><a href="http://www.cs.umd.edu/">UMD Computer Science</a></li>
		      <li><a href="http://www.math.berkeley.edu/">Berkeley Mathematics</a></li>
		      <li><a href="http://www.umiacs.umd.edu/">UMIACS</a></li>
		    </ul>
		  </div>        
          
    </div> <!-- End of "rightside" -->
    <div id="leftside">
  	  <table><tbody><tr><td cellpadding="7">
  	  <h1>Manual</h1><br>

      <div id="toc">
  	    <ul>
  	    <li><a href="#whis">What is TopHat?</a></li>
  	    <li><a href="#prer">Prerequisites</a></li>
  	    <li><a href="#toph">Using TopHat</a></li>

  	    </ul><br>
  	  </div>
  	  <h2 id="whis">What is TopHat?</h2><br>
  	  <p> TopHat is a program that aligns RNA-Seq reads to a genome in order 
		to identify exon-exon splice junctions.  It is built on the ultrafast
	  short read mapping program <a href="http://bowtie-bio.sourceforge.net">Bowtie</a>.
	  TopHat runs on <strong>Linux</strong> and <strong>OS X</strong>.
	  </p> 
  	  <br>
  	  <h2 id="whis">What types of reads can I use TopHat with?</h2><br>
  	  <p>TopHat was designed to work with reads produced by the Illumina Genome 
	  Analyzer, although users have been successful in using TopHat with reads
	  from other technologies.  In TopHat 1.1.0, we began supporting Applied Biosystems' Colorspace format.
	  The software is optimized for reads 75bp or longer.  
	  </p> 
	  <p>Mixing paired- and single- 
		end reads together is <strong>not</strong> supported.
	</p>
  	  <br>
	  <h2 id="whis">How does TopHat find junctions?</h2><br>
  	  <p> TopHat finds splice junctions without a reference annotation.  
By 
		first mapping RNA-Seq reads to the genome, TopHat identifies potential
		exons, since many RNA-Seq reads will contiguously align to the genome.
 Using this initial mapping information, TopHat builds a database of 
possible splice
		junctions, and then maps the reads against these junctions to confirm
		them.  
	  </p> 
	
      <p> Short read sequencing machines can currently produce reads 100bp or 
	  longer, but many exons are shorter than this, and so would be missed in
	  the initial mapping.  TopHat solves this problem by splitting all input
	  reads into smaller segments, and then mapping them independently.  The 
	  segment alignments are "glued" back together in a final step of the 
	  program to produce the end-to-end read alignments.
	  </p>
	
	  <p> TopHat generates its database of possible splice junctions from three
	  sources of evidence.  The first source is pairings of "coverage islands", 
	  which are distinct regions of piled up reads in the initial mapping.  
	  Neighboring islands are often spliced together in the transcriptome, so 
	  TopHat looks for ways to join these with an intron.  The second
	  source is only used when TopHat is run with paired end reads.  When reads 
	  in a pair come from different exons of a transcript, they will generally
	  be mapped far apart in the genome coordinate space.  When this happens,
	  TopHat tries to "close" the gap between them by looking for subsequences
	  of the genomic interval between mates with a total length about equal to
	  the expected distance between mates.  The "introns" in this subsequence are added to 
	  the database.  The third, and strongest, source of evidence for a splice
	  junction is when two segments from the same read are mapped far apart, or
	  when an internal segment fails to map. With long (&gt;=75bp) reads, "GT-AG", 
	  "GC-AG" and "AT-AC" introns will be found ab initio.  With shorter reads, 
	  TopHat only reports alignments across "GT-AG" introns</p>
	
  	  <br>
  	  <h2 id="bins">Prerequisites</h2><br>
	  <p> To use TopHat, you will need the following programs in your <tt>PATH</tt>:
      </p><ul>
         <li><tt>bowtie2 and bowtie2-align (or bowtie)<br>
</tt></li>
         <li><tt>bowtie2-inspect (or bowtie-inspect)<br>
</tt></li>
         <li><tt>bowtie2-build (or bowtie-build)</tt></li>
  <li><tt>samtools</tt></li>

      </ul>
	  <p>Because TopHat outputs alignments in BAM format, you will need to download and install the <a href="http://samtools.sourceforge.net/">SAM tools</a>. The Boost You may want to take a look at the <a href="tutorial.html">Getting started</a> guide for more detailed installation instructions, 
      including installation of SAM tools and Boost.
	  </p><p>You will also need Python <strong>version 2.4 or higher</strong>.
      </p>
  	  <br>
  	  <h2 id="bins">Obtaining and installing TopHat</h2><br>
	  <p> You can download the latest source release <a href="./downloads/">here</a>. </p>
	  <p> To install TopHat, unpack the tarball and change to the package 
		directory as follows: </p> <p>
		</p><blockquote>tar zxvf tophat-2.0.0.tar.gz</blockquote> 
		<blockquote>cd tophat-2.0.0/</blockquote>
	  <p>Now build the package: </p> <p> 
		</p><blockquote>./configure --prefix=&lt;install_prefix&gt;                       --with-boost=&lt;boost_install_prefix&gt;
		      --with-bam=&lt;samtools_install_prefix&gt;</blockquote>
		<blockquote>make</blockquote>
	  <p>Finally, install TopHat:</p><p>
		</p><blockquote>make install</blockquote>
  	  

	  <p>Below, you will find a detailed list of command-line options you can
		use to control TopHat.  Beginning users should take a look at the 
		<a href="tutorial.html">Getting started</a> guide for a tutorial on
		installing and running TopHat.

	  <br>
  	  
	  </p><div id="subheader">
	  <table width="100%"><tbody><tr> <td> <strong>Please Note</strong> TopHat has a number
		of parameters and options, and their default values are tuned for processing
		<strong>mammalian</strong> RNA-Seq reads. If you would like to use TopHat for
		another class of organism, we recommend setting some of the parameters 
		with more strict, conservative values than their defaults.  Usually, setting
		the maximum intron size to 4 or 5 Kb is sufficient to discover most 
		junctions while keeping the number of false positives low.

 		</td></tr></tbody></table> </div><br>
	  <h2 id="algn">Using <tt>TopHat</tt></h2><br>
  	  
	  <p>
	  The following is a detailed description of the options used to control
	  the <tt>tophat</tt> script:
	  </p><br>
	  <blockquote>
 	   Usage: tophat [options]* &lt;index_base&gt; &lt;reads1_1[,...,readsN_1]&gt; [reads1_2,...readsN_2] <br>
	  </blockquote>
		When running TopHat with paired ends, it is <strong>critical</strong> 
		that the *_1 files an the *_2 files appear in separate comma separated
		lists, and that the order of the files in the two lists is the same. 
	
	  <p><strong>NOTE:</strong> TopHat can align reads that are up to 1024 bp, 
		and it handles paired end reads, but we <strong>do not recommend</strong> 
		mixing several "types" of reads in the same TopHat run.  For example,
		mixing 100bp single end reads and 2x27bp paired ends into the same TopHat
		run will give bad results. If you'd like to combine results from several
		"flavors" of RNA-Seq reads, you can run first with one of your sets,
		and feed the junctions produced by that run into future TopHat runs as
		externally supplied junctions with the <tt>-j</tt> option (see below)</p>
	  <table cellspacing="15">
	  <tbody><tr><td valign="top">
	  <b>Arguments:</b>
	  </td><td valign="top">
	  <br>
</td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>&lt;ebwt_base&gt;</tt>
	  </td><td valign="top">
	  The basename of the index to be searched.  The
      basename is the name of any of the five index
      files up to but not including the first period.
      bowtie first looks in the current directory for
      the index files, then looks in the <tt>indexes</tt>
      subdirectory under the directory where the
      currently-running <tt>bowtie</tt> executable is located,
      then looks in the directory specified in the
      <tt>BOWTIE_INDEXES</tt> environment variable.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>&lt;reads1_1[,...,readsN_1]&gt;</tt>
	  </td><td valign="top">
	  A comma-separated list of files containing reads in FASTQ or FASTA format.  
	  When running TopHat with paired-end reads, this should be the *_1 ("left")
	  set of files.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>&lt;[reads1_2,...readsN_2]&gt;</tt>
	  </td><td valign="top">
	  A comma-separated list of files containing reads in FASTA or FASTA format.  
	  Only used when running TopHat with paired end reads, and contains the 
	  *_2 ("right") set of files.  The *_2 files <strong>MUST</strong> appear 
	  in the same order as the *_1 files.
	  </td></tr>
	  <tr><td valign="top">
	  <b>Options:</b>
	  </td><td valign="top">
	  <br>
</td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-h/--help</tt>
	  </td><td valign="top">
	  Prints the help message and exits   
	  </td></tr>
	 <tr><td nowrap="nowrap" valign="top">
	  <tt>-v/--version</tt>
	  </td><td valign="top">
	  Prints the TopHat version number and exits   
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--bowtie1</tt>
	  </td><td valign="top">
	  Uses Bowtie1 instead of Bowtie2.
	  If you use colorspace reads, you need to use this option
	  as Bowtie2 does not support colorspace reads.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-o/--output-dir &lt;string&gt;</tt>
	  </td><td valign="top">
	  Sets the name of the directory in which TopHat will write all of its 
	  output.  The default is "./tophat_out".
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-r/--mate-inner-dist &lt;int&gt;</tt>
	  </td><td valign="top">
	  This is the expected (mean) inner distance between mate pairs.  For, 
	  example, for paired end runs with fragments selected at 300bp, where each
	  end is 50bp, you should set -r to be 200.  There is no default, and this
	  parameter is required for paired end runs.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--mate-std-dev &lt;int&gt;</tt>
	  </td><td valign="top">
	  The standard deviation for the distribution on inner distances between
	  mate pairs.  The default is 20bp.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-a/--min-anchor-length &lt;int&gt;</tt>
	  </td><td valign="top">
	  The "anchor length".  TopHat will report junctions spanned by reads
	  with at least this many bases on each side of the junction. Note that
	  individual spliced alignments may span a junction with fewer than this 
	  many bases on one side.  However, every junction involved in spliced 
	  alignments is supported by at least one read with this many bases on each
	  side.	  This must be at least 3 and the default is 8.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-m/--splice-mismatches &lt;int&gt;</tt>
	  </td><td valign="top">
	  The maximum number of mismatches that may appear in the "anchor" region
	  of a spliced alignment.  The default is 0.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-i/--min-intron-length &lt;int&gt;</tt>
	  </td><td valign="top">
	  The minimum intron length.  TopHat will ignore donor/acceptor pairs 
	  closer than this many bases apart.  The default is 70.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-I/--max-intron-length &lt;int&gt;</tt>
	  </td><td valign="top">
	  The maximum intron length.  When searching for junctions ab initio, 
	  TopHat will ignore donor/acceptor pairs farther than this many bases 
	  apart, except when such a pair is supported by a split segment alignment
	  of a long read.  The default is 500000.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--max-insertion-length &lt;int&gt;</tt>
	  </td><td valign="top">
	  The maximum insertion length.  The default is 3.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--max-deletion-length &lt;int&gt;</tt>
	  </td><td valign="top">
	  The maximum deletion length.  The default is 3.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--solexa-quals</tt>
	  </td><td valign="top">
	  Use the Solexa scale for quality values in FASTQ files.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--solexa1.3-quals</tt>
	  </td><td valign="top">
	  As of the Illumina GA pipeline version 1.3, quality scores are encoded 
	  in Phred-scaled base-64.  Use this option for FASTQ files from pipeline 1.3 or later.  
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-Q/--quals</tt>
	  </td><td valign="top">
	  Separate quality value files - colorspace read files (CSFASTA) come with separate qual files.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--integer-quals</tt>
	  </td><td valign="top">
	  Quality values are space-delimited integer values, this becomes default when you specify -C/--color.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-C/--color</tt>
	  </td><td valign="top">
	  Colorspace reads, note that it uses a colorspace bowtie index and requires Bowtie 0.12.6 or higher.<br>
	  Common usage: tophat --color --quals [other options]* 
&lt;colorspace_index_base&gt; &lt;reads1_1[,...,readsN_1]&gt; 
[reads1_2,...readsN_2] &lt;quals1_1[,...,qualsN_1]&gt; 
[quals1_2,...qualsN_2]
	  </td></tr>
      <!--
	  <tr><td VALIGN=top nowrap>
	  <tt>-F/--min-isoform-fraction &lt;0.0-1.0&gt;</tt>
	  </td><td VALIGN=top>
	  TopHat filters out junctions supported by too few alignments.  Suppose 
	  a junction spanning two exons, is supported by S reads.  Let the average
	  depth of coverage of exon A be D, and assume that it is higher than B.  
	  If S / D is less than the minimum isoform fraction, the junction is 
	  not reported.  A value of zero disables the filter. The default is 0.15.
	  </td></tr>
      -->
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-p/--num-threads &lt;int&gt;</tt>
	  </td><td valign="top">
	  Use this many threads to align reads. The default is 1.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-g/--max-multihits &lt;int&gt;</tt>
	  </td><td valign="top">
	  Instructs TopHat to allow up to this many alignments to the reference
	  for a given read, and suppresses all alignments for reads with more than
	  this many alignments. The default is 20 for read mapping.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--report-secondary-hits</tt>
	  </td><td valign="top">
	      Without the option, TopHat will report best or primary alignments based on alignment scores (AS).
	      If you want to output additional or secondary alignments, use the option,
	      which will report up to 20 alignments
	      (the default number 20 can be changed using -g/--max-multihits option above).
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--report-discordant-pair-alignments</tt>
	  </td><td valign="top">
	      This option will allow mate pairs to map to different chromosomes,
	      distant places on the same chromosome, or on the same strand.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--no-closure-search</tt>
	  </td><td valign="top">
	  Disables the mate pair closure-based search for junctions.  Currently, 
	  has no effect - closure search is off by default.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--closure-search</tt>
	  </td><td valign="top">
	  Enables the mate pair closure-based search for junctions.  Closure-based
	  search should only be used when the expected inner distance between mates
	  is small (&lt;= 50bp)
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--no-coverage-search</tt>
	  </td><td valign="top">
	  Disables the coverage based search for junctions.  
	  </td></tr> 
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--coverage-search</tt>
	  </td><td valign="top">
	  Enables the coverage based search for junctions.  Use when coverage search
	  is disabled by default (such as for reads 75bp or longer), for maximum sensitivity.   
	</td></tr>
	  <tr><td valign="top">
	  <tt>--microexon-search</tt>
	  </td><td valign="top">
	  With this option, the pipeline will attempt to find alignments incident 
	  to microexons.  Works only for reads 50bp or longer.
	  </td></tr>
	  <tr><td valign="top">
	  <tt>--butterfly-search</tt>
	  </td><td valign="top">
	  TopHat will use a slower but potentially more sensitive algorithm to find
	  junctions <strong>in addition</strong> to its standard search.  Consider
	  using this if you expect that your experiment produced a lot of reads from 
	  pre-mRNA, that fall within the introns of your transcripts.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--library-type</tt>
	  </td><td valign="top">
		TopHat will treat the reads as strand specific.  Every read alignment will have an <tt>XS</tt> attribute tag.  Consider supplying library type options below to select the correct RNA-seq protocol.
		</td></tr>
	  </tbody></table>
	  <table cellspacing="15">
	    <tbody><tr>
	      <td nowrap="nowrap" valign="top"><strong>Library Type</strong></td><td nowrap="nowrap" valign="top"><strong>Examples</strong></td><td valign="top"><strong>Description</strong></td>
	    </tr>
	    <tr><td nowrap="nowrap" valign="top">fr-unstranded</td><td nowrap="nowrap" valign="top"><tt>Standard Illumina</tt></td><td valign="top">Reads
 from the left-most end of the fragment (in transcript coordinates) map 
to the transcript strand, and the right-most end maps to the opposite 
strand.</td></tr>
	    <tr><td nowrap="nowrap" valign="top">fr-firststrand</td><td nowrap="nowrap" valign="top"><tt>dUTP, NSR, NNSR</tt></td><td valign="top">Same
 as above except we enforce the rule that the right-most end of the 
fragment (in transcript coordinates) is the first sequenced (or only 
sequenced for single-end reads).  Equivalently, it is assumed that only 
the strand generated during first strand synthesis is sequenced.</td></tr>
	    <tr><td nowrap="nowrap" valign="top">fr-secondstrand</td><td nowrap="nowrap" valign="top"><tt>Ligation, Standard SOLiD</tt></td><td valign="top">Same
 as above except we enforce the rule that the left-most end of the 
fragment (in transcript coordinates) is the first sequenced (or only 
sequenced for single-end reads).  Equivalently, it is assumed that only 
the strand generated during second strand synthesis is sequenced.</td></tr>
	  </tbody></table>
	</td></tr></tbody></table>
	<table><tbody><tr><td cellpadding="7">  
	  <table cellspacing="15">
	  <tbody><tr><td valign="top">      
	  <b>Advanced Options:</b>
	  </td><td valign="top">
	  <br>
</td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-n/--transcriptome-mismatches</tt>
	  </td><td valign="top">
	  Maximum number of mismatches allowed when reads are aligned to the 
transcriptome.  The default is 2.
When Bowtie2 is used, this number is also used to decide whether or not
	  to further re-align some of the transcriptome-mapped reads to the 
genome.	If the alignment score of the best alignment among multiple 
candidates for a read is lower than "bowtie2-min-score",
	  which is internally defined as (<i>max_penalty</i> - 1) * <i>max_mismatches</i>, then the reads will be kept for re-alignment through the rest of the pipeline.&nbsp;  You can specify <i>max_penalty</i> via "--b2-mp" option.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--genome-read-mismatches</tt>
	  </td><td valign="top">
	  When whole reads are first mapped on the genome, this many mismatches
	  in each read alignment are allowed.  The default is 2.
	  This number is also used to decide whether to further re-align some 
of the reads (by splitting them into segments) with a similar scoring 
threshold scheme as described for the --transcriptome-mismatches option 
above.<br>
</td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--read-mismatches</tt>
	  </td><td valign="top">
	  Final read alignments having more than these many mismatches are discarded.
	  The default is 2.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--bowtie-n</tt>
	  </td><td valign="top">
	  TopHat uses "-v" in Bowtie for initial read mapping (the default), but with this option, "-n" is used instead.
	  Read segments are always mapped using "-v" option.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--segment-mismatches</tt>
	  </td><td valign="top">
	  Read segments are mapped independently, allowing up to this many mismatches
	  in each segment alignment.  The default is 2.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--segment-length</tt>
	  </td><td valign="top">
	  Each read is cut up into segments, each at least this long.  These segments
	  are mapped independently. The default is 25.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--min-closure-exon</tt>
	  </td><td valign="top">
	  During closure search for paired end reads, exonic hops in the potential 
	  splice graph must be at least this long.  The default is 50.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--min-closure-intron</tt>
	  </td><td valign="top">
	  The minimum intron length that may be found during closure search.  
	  The default is 50.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--max-closure-intron</tt>
	  </td><td valign="top">
	  The maximum intron length that may be found during closure search.  
	  The default is 5000.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--min-coverage-intron</tt>
	  </td><td valign="top">
	  The minimum intron length that may be found during coverage search.  
	  The default is 50.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--max-coverage-intron</tt>
	  </td><td valign="top">
	  The maximum intron length that may be found during coverage search.  
	  The default is 20000.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--min-segment-intron</tt>
	  </td><td valign="top">
	  The minimum intron length that may be found during split-segment search.  
	  The default is 50.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--max-segment-intron</tt>
	  </td><td valign="top">
	  The maximum intron length that may be found during split-segment search.  
	  The default is 500000.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--keep-tmp</tt>
	  </td><td valign="top">
	  Causes TopHat to preserve its intermediate files produced during the 
run (mostly useful for debugging). The default is to delete these 
temporary files.<br>
</td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--keep-fasta-odrer</tt>
	  </td><td valign="top">
	      In order to sort alignments in the same order in the genome fasta file,
	      the option can be used.
	      But this option will make the output SAM/BAM file incompatible with
	      those from the previous versions of TopHat (1.4.1 or lower).
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--no-sort-bam</tt>
	  </td><td valign="top">
	  Output BAM is not coordinate-sorted.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--no-convert-bam</tt>
	  </td><td valign="top">
	  Do not convert to bam format. Output is &lt;output_dir&gt;/accepted_hit.sam.
	  Implies --no-sort-bam.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-z/--zpacker </tt>
	  </td><td valign="top">
	  Manually specify the program used for compression of temporary files;
	  default is <tt>gzip</tt>; use <tt>-z0</tt> to disable compression altogether. 
	  Any program that is option-compatible with gzip can be used (e.g. bzip2, pigz, 
	  pbzip2).
	  </td></tr>
	
	 </tbody></table><br>

	  <b>Bowtie 2 specific options:</b>
	  <a name="b2-option"></a>
	  <p> Bowtie 2 provides many options so that users can have more 
flexibility as to how reads are mapped. TopHat 2 allows users to pass many 
of these options to Bowtie 2 by preceding the Bowtie 2 
option name with the<b> <strong>--b2-</strong></b>
prefix.&nbsp;  Please refer to the <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie2</a> website for detailed information.
	  </p><table cellspacing="12" width="100%">
	  <tbody><tr>
              <td valign="top">
	      <b>Preset options </b>in --end-to-end mode&nbsp; (local alignment is not used in TopHat2):
	      <table cellspacing="10">
		<tbody><tr>
    <td valign="top"><i>Tophat 2 option:</i><br>
    </td>
    <td valign="top"><i>Corresponding Bowtie 2 option:</i><br>
    </td>
  </tr>
<tr><td nowrap="nowrap" valign="top">
		    <tt>--b2-very-fast</tt>
		</td><td valign="top">--very-fast<br>
</td>
</tr>
		<tr><td nowrap="nowrap" valign="top">
		    <tt>--b2-fast</tt>
		</td><td valign="top">--fast<br>
</td>
</tr>
		<tr><td nowrap="nowrap" valign="top">
		    <tt>--b2-sensitive</tt>
		</td><td valign="top">--sensitive<br>
</td>
</tr>
		<tr><td nowrap="nowrap" valign="top">
		    <tt>--b2-very-sensitive</tt>
		</td><td valign="top">--very-sensitive<br>
  </td>
  </tr>
	      </tbody></table>
 </td>
</tr><tr>  <td valign="top">
<b>Alignment options:
</b><table cellspacing="10">
<tbody><tr><td nowrap="nowrap" valign="top">
<tt>--b2-N</tt>
</td><td>
The default is 0.
</td></tr>
<tr><td nowrap="nowrap" valign="top">
<tt>--b2-L</tt>
</td><td>
The default is 20.
</td></tr>
<tr><td nowrap="nowrap" valign="top">
<tt>--b2-i</tt>
</td><td>
The default is S,1,1.25.
</td></tr>
<tr><td nowrap="nowrap" valign="top">
<tt>--b2-n-ceil</tt>
</td><td>
The default is L,0,0.15.
</td></tr>
<tr><td nowrap="nowrap" valign="top">
<tt>--b2-gbar</tt>
</td><td>
The default is 4.
</td></tr>
</tbody></table>
  </td>
</tr>
<tr><td valign="top">
	      <b>Scoring options:
	      </b><table border="0" cellspacing="10">
		<tbody><tr><td nowrap="nowrap" valign="top">
		    <tt>--b2-mp</tt>
		  </td><td>
		    The default is 6,2.
		</td></tr>
		<tr><td nowrap="nowrap" valign="top">
		    <tt>--b2-np</tt>
		  </td><td>
		    The default is 1.
		</td></tr>
		<tr><td nowrap="nowrap" valign="top">
		    <tt>--b2-rdg</tt>
		  </td><td>
		    The default is 5,3.
		</td></tr>
		<tr><td nowrap="nowrap" valign="top">
		    <tt>--b2-rfg</tt>
		  </td><td>
		    The default is 5,3.
		</td></tr>
		<tr><td nowrap="nowrap" valign="top">
		    <tt>--b2-score-min</tt>
		  </td><td>
		    The default is L,-0.6,-0.6.
		</td></tr>
	      </tbody></table>
	  </td>
</tr><tr>
<td valign="top">
          <b>Effort options:</b><br>

  <table cellspacing="10">

		<tbody><tr><td nowrap="nowrap" valign="top">
		    <tt>--b2-D</tt>
		  </td><td>
		    The default is 15.
		</td></tr>
		<tr><td nowrap="nowrap" valign="top">
		    <tt>--b2-R</tt>
		  </td><td>
		    The default is 2.
		</td></tr>
	      </tbody>
  </table>

</td>
</tr>
	  
	  </tbody></table>

	  <b>Fusion mapping options:</b><a name="fusion"></a>
	  <p> Reads can be aligned to potential fusion transcripts if the --fusion-search option is specified.
	    The fusion alignments are reported in SAM format using custom fields XF and XP (see the output format)
	    and some additional information about fusions will be reported (see fusions.out).
	    Once mapping is done, you can run tophat-fusion-post to filter out fusion transcripts
	    (see the TopHat-Fusion website for more details).
	  </p><table cellspacing="15">
	    <tbody><tr><td nowrap="nowrap" valign="top">
		<tt>--fusion-search</tt>
	      </td><td valign="top">
		Turn on fusion mapping
	    </td></tr>
	    <tr><td nowrap="nowrap" valign="top">
		<tt>--fusion-anchor-length</tt>
	      </td><td valign="top">
		A "supporting" read must map to both sides of a fusion by at least this many bases. The default is 20.
	    </td></tr>
	    <tr><td nowrap="nowrap" valign="top">
		<tt>--fusion-min-dist</tt>
	      </td><td valign="top">
		For intra-chromosomal fusions, TopHat-Fusion tries to find fusions separated by at least this distance.
		The default is 10000000. 
	    </td></tr>
	    <tr><td nowrap="nowrap" valign="top">
		<tt>--fusion-read-mismatches</tt>
	      </td><td valign="top">
		Reads support fusions if they map across fusion with at most this many mismatches. The default is 2. 
	    </td></tr>
	    <tr><td nowrap="nowrap" valign="top">
		<tt>--fusion-multireads</tt>
	      </td><td valign="top">
		Reads that map to more than this many places will be ignored.
		It may be possible that a fusion is supported by reads (or pairs) that map to multiple places.
		The default is 2.
	    </td></tr>
	    <tr><td nowrap="nowrap" valign="top">
		<tt>--fusion-multipairs</tt>
	      </td><td valign="top">
		Pairs that map to more than this many places will be ignored. The default is 2.
	    </td></tr>
	      <tr><td nowrap="nowrap" valign="top">
		<tt>--fusion-ignore-chromosomes</tt>
	      </td><td valign="top">
		  Ignore some chromosomes such as chrM when detecting fusion break points.
		  Please check the correct names for chromosomes,
		  that is, mitochondrial DNA is represented as chrM or M depending on the annotation you use.
	    </td></tr>
	  </tbody></table>

	<b>Supplying your own transcript annotation data:</b>
	  <p>The options below allow you validate your own list of known transcripts or junctions with your 
		RNA-Seq data. Note that the chromosome names in the files provided
		with the options below <strong>must match</strong> the names in the 
		Bowtie index.  These names are <strong>case-senstitive</strong>.</p>
	  <table cellspacing="15">
	  <tbody><tr><td valign="top">
	  <br>
</td><td valign="top">
	  <br>
</td></tr>
	 <tr>
	  <td nowrap="nowrap" valign="top">
	    <tt>-j/--raw-juncs &lt;.juncs file&gt;</tt>
	  </td><td valign="top">
	  <p> Supply TopHat with a list of raw junctions.  Junctions are specified 
		one per line, in a tab-delimited format.  Records look like:
	  </p> <br>
		<tt>&lt;chrom&gt;   &lt;left&gt;   &lt;right&gt;   &lt;+/-&gt;</tt>
		<p><tt>left</tt> and <tt>right</tt> are zero-based coordinates, and 
		  specify the last character of the left sequenced to be spliced to 
		  the first character of the right sequence, inclusive.
		  That is, the last and the first positions of the flanking exons.
		  Users can convert <tt>junctions.bed</tt> (one of the TopHat outputs) to this format
		  using <strong>bed_to_juncs &lt; junctions.bed &gt; new_list.juncs</strong>
		  where bed_to_juncs can be found under the same folder as tophat</p>
	  </td></tr>

	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--no-novel-juncs</tt>
	  </td><td valign="top">
	  Only look for reads across junctions indicated in the supplied GFF or junctions file. (ignored without -G/-j)
	  </td></tr>

	  <tr><td nowrap="nowrap" valign="top">
	  <a name="transcriptome"></a>
	  <tt>-G/--GTF &lt;GTF/GFF3 file&gt;</tt>
	  </td><td valign="top">
	  <p> Supply TopHat with a set of gene model annotations and/or known transcripts, as a GTF 2.2 or GFF3 formatted file. 
	  If this option is provided, TopHat will first extract the transcript sequences and use Bowtie to align reads 
	  to this virtual transcriptome first. Only the reads that do not fully map to the transcriptome will then be mapped 
	  on the genome. The reads that did map on the transcriptome will be converted to genomic mappings (spliced as needed) 
	  and merged with the novel mappings and junctions in the final tophat output.</p>
	 <p>Please note that the values in the first column of the provided GTF/GFF file (column which 
		indicates the chromosome or contig on which the feature is located),  <strong>must match</strong> the 
		name of the reference sequence in the Bowtie index you are using with TopHat.  You can get a list of the sequence names
		in a Bowtie index by typing:</p>
		<br>
	    	<blockquote>
				bowtie-inspect --names your_index
			</blockquote> 
		<br>
	  So before using a known annotation file with this option please make sure that the 1st column in the annotation file
	  uses the <strong>exact same chromosome/contig names (case sensitive)</strong> as shown by the <tt>bowtie-inspect</tt> command above.
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <a name="t_idx"></a>
	  <tt>--transcriptome-index &lt;dir/prefix&gt;</tt>
	  </td><td valign="top">
	  <p> When providing TopHat with a known transcript file (-G/--GTF 
option above), a transcriptome sequence file is built 
	  and a Bowtie index has to be created for it in order to align the 
reads to the known transcripts. Creating this Bowtie index 
	  can be time consuming and in many cases the same transcriptome data 
is being used for aligning multiple samples with TopHat.
	  A transcriptome index and the associated data files (the original GFF
 file) can be thus reused for multiple TopHat runs with this option, 
	  so these files are only created for the first run with a given set of
 transcripts. If multiple TopHat runs are planned with the same 
transcriptome data, 
	  TopHat should be first run with the -G option and with the 
--transcriptome-index option pointing to a directory and a name prefix 
which will indicate 
	  where the transcriptome data files will be stored. Then subsequent 
TopHat runs using the same --transcriptome-index option value will 
directly use
	  the transcriptome data created in the first run (no -G option needed 
for subsequent runs).</p>
	  For example the first TopHat run could look like this:
	  <br>
	  <tt>tophat -o out_sample1 -G known_genes.gtf \<br>
      --transcriptome-index=transcriptome_data/known \<br>
      hg19 sample1_1.fq.z</tt>
      <br>
	  In this example the first run will create the transcriptome_data 
directory if it doesn't exist, and files known.fa, known.gff and 
known.*ebwt 
	  (Bowtie index files) will be generated in that directory. Then for 
subsequent runs with the same genome and known transcripts but different
 reads 
	  (e.g. sample2_2.fq.z etc.), TopHat will no longer spend time building
 the transcriptome index because it can directly use the previously 
built 
	  transcriptome index, so the -G option can be even discarded for 
subsequent runs:
	  <br>
	  <tt>tophat -o out_sample2 \<br>
      --transcriptome-index=transcriptome_data/known \<br>
      hg19 sample2_1.fq.z</tt>
      <br>
	  </td></tr>
	  <tr><td colspan="2">
	  (The following options in this section are only used when the transcriptome search was activated with <tt>-G/--GTF</tt> or <tt>--transcriptome-index</tt>)
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-T/--transcriptome-only</tt>
	  </td><td valign="top">
	  Only align the reads to the transcriptome and report only those mappings as genomic mappings.
	  </td></tr>

	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-x/--transcriptome-max-hits</tt>
	  </td><td valign="top">
	  Maximum number of mappings allowed for a read, when aligned to the 
transcriptome (any reads found with more then this number of mappings 
	  will be discarded).
	  </td></tr>

	  <tr><td nowrap="nowrap" valign="top">
	  <tt>-M/--prefilter-multihits</tt>
	  </td><td valign="top">
	  When mapping reads on the transcriptome, some repetitive or low 
complexity reads that would be discarded in the context 
      of the genome may appear to align to the transcript sequences and 
thus may end up reported as mapped to those genes only. This option 
directs TopHat to 
      first align the reads to the whole genome in order to determine 
and exclude such multi-mapped reads (according to the value of the 
      <tt>-g/--max-multihits</tt> option).
	  </td></tr>
	   
	  </tbody></table>
	
	<b>Supplying your own insertions/deletions:</b>
	  <p>The options below allow you validate your own indels with your 
		RNA-Seq data.  Note that the chromosome names in the files provided
		with the options below <strong>must match</strong> the names in the 
		Bowtie index.  These names are <strong>case-senstitive</strong>.</p>
	  <table cellspacing="15">
	  <tbody><tr><td valign="top">

	  <br>
</td><td valign="top">
	  <br>
</td></tr>
	 <tr><td nowrap="nowrap" valign="top">
	  <tt>--insertions/--deletions &lt;.juncs file&gt;</tt>
	  </td><td valign="top">
	  <p> Supply TopHat with a list of insertions or deletions with respect to the reference.  Indels are specified 
		one per line, in a tab-delimited format, identical to that of junctions.  Records look like:
	  </p> <br>
		<tt>&lt;chrom&gt;   &lt;left&gt;   &lt;right&gt;   &lt;+/-&gt;</tt>
		<p><tt>left</tt> and <tt>right</tt> are zero-based coordinates, and 
			specify the last character of the left sequenced to be spliced to 
			the first character of the right sequence, inclusive. </p>
	  </td></tr>
	  <tr><td nowrap="nowrap" valign="top">
	  <tt>--no-novel-indels</tt>
	  </td><td valign="top">
	  Only look for reads across indels in the supplied indel file, or disable indel detection when no file has been provided.
	  </td></tr>

	  </tbody></table>
	
  	  <h2 id="output">TopHat Output</h2><br>
  	  <p>
  	  The <tt>tophat</tt> script produces a number of files in the directory
	  in which it was invoked.  Most of these files are internal, intermediate 
	  files that are generated for use within the pipeline.  The output 
	  files you will likely want to look at are:
      </p> <br>
      <ol>
	<li><tt>accepted_hits.bam</tt>. A list of read alignments in <a href="http://samtools.sourceforge.net">SAM</a> format.  SAM is a 
			compact short read alignment format that is increasingly being 
			adopted.  The formal specification is <a href="http://samtools.sourceforge.net/SAM1.pdf">here</a>.</li>
        <li><tt>junctions.bed</tt>.  A <a href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1">UCSC 
		BED</a> track of junctions reported by TopHat.  Each junction consists 
		of two connected BED blocks, where each block is as long as the 
		maximal overhang of any read spanning the junction.  The score is the 
		number of alignments spanning the junction.</li>
	<li><tt>insertions.bed and deletions.bed</tt>.  <a href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1">UCSC 
		BED</a> tracks of insertions and deletions reported by TopHat.
	  <br>Insertions.bed - chromLeft refers to the last genomic base before the insertion.
	  <br>Deletions.bed - chromLeft refers to the first genomic base of the deletion.
	  </li>
      </ol>
      </td></tr></tbody></table><br>
    </div>
  </div>
  <div id="footer">
  	<table cellspacing="15" width="100%"><tbody><tr><td>
    This research was supported in part by NIH grants R01-LM06845 and R01-GM083873 and NSF grant CCF-0347992.
    </td><td align="right">
    Administrator: <a href="mailto:cole@cs.umd.edu">Cole Trapnell</a>. Design by <a href="http://www.free-css-templates.com" title="Design by David Herreman">David Herreman</a>
    </td></tr></tbody></table>
  </div>
</div>



<!-- Google analytics code -->
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
var pageTracker = _gat._getTracker("UA-6101038-1");
pageTracker._trackPageview();
</script>

</body></html>
