<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>TopHat :: Center for Bioinformatics and Computational Biology</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<link rel="stylesheet" type="text/css" href="css/style.css" media="screen" />
</head>
<body>
<div id="wrap">
  <div id="top">
    <div class="lefts">
    <table width="100%" cellpadding="2">
      <tr><td>
           <a href="./index.html"><h1>TopHat</h1></a>
              <h2>A spliced read mapper for RNA-Seq</h2>
      </td><td align="right" valign="middle">
        <a href="http://bio.math.berkeley.edu/">
        <img style="vertical-align:middle;padding-top:4px" 
          src="images/UCBerkeley-seal.scaled.gif" border="0">
        </a>&nbsp;
        <a href="http://genomics.jhu.edu/">
        <img style="vertical-align:middle;padding-top:4px"
            src="images/JHU-seal.gif" border="0">
        </a>&nbsp;
        <a href="http://www.cbcb.umd.edu/"><img style="vertical-align:middle;padding-top:4px" border=0 src="images/cbcb_logo.gif"></a>&nbsp;&nbsp;
      </td></tr>
      </td></tr>
    </table>
    </div>
  </div>

  <div id="main">
	
   <div id="rightside">

		<h2>Site Map</h2>
		  <div class="box">
		    <ul>
		      <li><a href="index.html">Home</a></li>
		      <li><a href="tutorial.html">Getting started</a></li>
		      <li><a href="manual.html">Manual</a></li>
		      <li><a href="igenomes.html">Index and annotation downloads</a></li>
		    </ul>
		  </div>

		  <h2><u>News and updates</u></h2>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tr><td>New releases and related tools will be announced through the <a href="https://lists.sourceforge.net/lists/listinfo/bowtie-bio-announce"><b>mailing list</b></a></td></tr>
			  </table>
		    </ul>
		  </div>

		  <h2><u>Getting Help</u></h2>
		  <div class="box">
		    <ul>
		      <table width="100%">
				<tr><td>Questions about TopHat should be sent to <a href="mailto:tophat.cufflinks@gmail.com"><b>tophat.cufflinks@gmail.com</b></a>.  Please do not email technical questions to TopHat contributors directly.</td></tr>
			  </table>
		    </ul>
		  </div>

		  <a href="./downloads"><h2><u>Releases</u></h2></a>
		  <div class="box">
		    <ul>
		      <table width="100%"><tr>
		      <td>
				<tr><td>version 1.4.0 (BETA)</td> <td align="right">1/5/2012</td></tr>
				<tr>
		      		<td><a href="./downloads/tophat-1.4.0.tar.gz" onClick="javascript: pageTracker._trackPageview('/downloads/tophat'); ">&nbsp;&nbsp;&nbsp;Source code</a></td>
		      	</tr>
				<tr>
					<td><a href="./downloads/tophat-1.4.0.Linux_x86_64.tar.gz" onClick="javascript: pageTracker._trackPageview('/downloads/tophat'); ">&nbsp;&nbsp;&nbsp;Linux x86_64 binary</a></td>
				</tr>
				<tr>
					<td><a href="./downloads/tophat-1.4.0.OSX_x86_64.tar.gz" onClick="javascript: pageTracker._trackPageview('/downloads/tophat'); ">&nbsp;&nbsp;&nbsp;Mac OS X x86_64 binary</a></td>
				</tr>
		      </td>
		      </tr>
			  </table>
		    </ul>
		  </div>

		  <h2>Related Tools</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://cufflinks.cbcb.umd.edu/">Cufflinks</a>: Isoform assembly and quantitation for RNA-Seq</li>
		      <li><a href="http://bowtie.cbcb.umd.edu">Bowtie</a>: Ultrafast short read alignment</li>
		      <li><a href="http://tophat-fusion.sourceforge.net">TopHat-Fusion</a>: An algorithm for Discovery of Novel Fusion Transcripts</li>
		    </ul>
		  </div>

		  <h2>Publications</h2>
		  <div class="box">
		    <ul>
			  <li style="font-size: x-small; line-height: 130%"><p>Trapnell C, Pachter L, Salzberg SL. <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/btp120"><b>TopHat: discovering splice junctions with RNA-Seq</b></a>. <i><a href="http://bioinformatics.oxfordjournals.org">Bioinformatics</a></i>  doi:10.1093/bioinformatics/btp120</p><br/></li>
		      <li style="font-size: x-small; line-height: 130%"><p>Langmead B, Trapnell C, Pop M, Salzberg SL. <a href="http://genomebiology.com/2009/10/3/R25"><b>Ultrafast and memory-efficient alignment of short DNA sequences to the human genome</b></a>. <i><a href="http://genomebiology.com">Genome Biology</a></i> <b>10</b>:R25.</li>
		    </ul>
		  </div>

		  <h2>Contributors</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://www.cs.umd.edu/~cole/">Cole Trapnell</a></li>
			  <li><a href="http://www.cs.umd.edu/~infphilo/">Daehwan Kim</a></li>
			  <li>Geo Pertea</li>
			  <li><a href="http://cs.berkeley.edu/~pimentel">Harold Pimentel</a></li>
			  <li>Ryan Kelley</li>
		      <li><a href="http://www.math.berkeley.edu/~lpachter/">Lior Pachter</a></li>
		      <li><a href="http://www.cbcb.umd.edu/~salzberg/">Steven Salzberg</a></li>
		    </ul>
		  </div>

		  <h2>Links</h2>
		  <div class="box">
		    <ul>
		      <li><a href="http://www.umd.edu/">University of Maryland</a></li>
		      <li><a href="http://www.berkeley.edu/">University of California, Berkeley</a></li>
		      <li><a href="http://www.cbcb.umd.edu/">UMD CBCB</a></li>
		      <li><a href="http://www.cs.umd.edu/">UMD Computer Science</a></li>
		      <li><a href="http://www.math.berkeley.edu/">Berkeley Mathematics</a></li>
		      <li><a href="http://www.umiacs.umd.edu/">UMIACS</a></li>
		    </ul>
		  </div>

    </div> <!-- End of "rightside" -->
    <div id="leftside">
  	<table><tr><td cellpadding=7>
  	<h1> Getting started</h1><br/>
      <div id="toc">
  	    <ul>
  	    <li><a href="#inst">Install quick-start</a></li>
		<li><a href="#ref">Test the installation</a></li>
		<li><a href="#ref">Preparing your reference</a></li>
  	    <li><a href="#reads">Preparing your reads</a></li>
  	    <li><a href="#toph">Running TopHat</a></li>
  	    <li><a href="#exam">Examining your results</a></li>
  	    </ul><br/>
  	  </div>
	  	
	<br/>
  	<h2 id="inst">Install quick-start</h2><br/>
      	<p>Download and extract the latest <a href="./downloads/">TopHat</a>
            and <a href="https://sourceforge.net/project/showfiles.php?group_id=236897&package_id=288231">Bowtie</a> releases.</p>
            You will need the following Bowtie programs in your <tt>PATH</tt>:
            <ul>
        	      <li><tt>bowtie</tt></li>
        	      <li><tt>bowtie-build</tt></li>
				  <li><tt>bowtie-inspect</tt></li>
        	</ul>
            You will also need the <tt>samtools</tt> program in your <tt>PATH</tt>, and if you plan to compile Tophat from source 
            you should also install the SAM tools package, by following these steps:
            <ul><ul>
               <li><a href="http://samtools.sourceforge.net/">Download</a>
                  the SAM tools</li>
                <li>Unpack the SAM tools tarball and <tt>cd</tt> to
                  the SAM tools source directory.</li>
                <li>Build the SAM tools by typing <tt>make</tt> at
                  the command line.</li>
                <li>Choose a directory into which you wish to copy
                  the SAM tools binary, the included library <tt>libbam.a</tt>,
                  and the library headers. A common choice is <tt>/usr/local/</tt>.
                </li>
                <li>Copy libbam.a to the <tt>lib/</tt> directory in
                  the folder you've chosen above (e.g. <tt>/usr/local/lib/</tt>)</li>
                <li>Create a directory called "<tt>bam</tt>" in the
                  <tt>include/</tt> directory (e.g.
                  /usr/local/include/bam) </li>
                <li>Copy the headers (files ending in .h) to the <tt>include/bam</tt>
                  directory you've created above (e.g. <tt>/usr/local/include/bam</tt>)
                </li>
                <li>Copy the samtools binary to some directory in
                      your PATH.</li>
            </ul></ul>
      	<p>
		<p> To install TopHat, unpack the tarball and change to the package 
			directory as follows: </p> <p>
			<blockquote>tar zxvf tophat-1.2.0.tar.gz</blockquote> 
			<blockquote>cd tophat-1.2.0/</blockquote>
		  <p>Now build the package: </p> <p> 
			<blockquote>./configure --prefix=/path/to/install/directory/</blockquote>
            <ul><ul>
            <p><b>Note:</b> 
            If you copied the SAM tools library and header files to someplace other than <tt>/usr/local/</tt>, you
            may need to supply the <tt>--with-bam</tt> configuration option. 
            For example, if you placed the libbam.a files in <tt>/opt/biotools/lib/</tt> and SAM tools .h files in <tt>/opt/biotools/include/bam/</tt>
            then the <tt>configure</tt> command will become something like this:
            <blockquote>./configure --prefix=/opt/biotools --with-bam=/opt/biotools</blockquote>
            (in this example tophat and its modules will be installed in <tt>/opt/biotools/bin/</tt> directory)
            <p>
            </ul></ul>
            After the <tt>configure</tt> script finishes successfully, build the Tophat package:
			<blockquote>make</blockquote>
		  <p>Finally, install TopHat:</p><p>
			<blockquote>make install</blockquote>
			<p>This will install tophat and its modules into <tt>/path/to/install/directory/bin</tt> directory. 
            You may want to add that directory to your shell's <tt>PATH</tt> if it's not there already.<p>
      	<br/>

		<h2 id="test">Testing the installation</h2><br/>
		<p>
		After you installed Bowtie, Samtools and TopHat, you should 
		test the pipeline on a simple test data set, which you can download 
		<a href="./downloads/test_data.tar.gz">here</a>.  This data is not meant
		to exhaustively test all the features of TopHat.  It's just to 
		verify that the installation worked.  Unzip the data, change 
		to the test_data directory and then run tophat:
		</p>  		
		<br/>
		<blockquote>tar zxvf test_data.tar.gz</blockquote>
		<blockquote>cd test_data</blockquote>
		<blockquote>tophat -r 20 test_ref reads_1.fq reads_2.fq</blockquote>
		<br/>
		<p>
		If TopHat ran successfully, you should see some lines of 
		output, like this:
		</p>
		<br/>
		<pre>
			
[Mon May  4 11:07:23 2009] Beginning TopHat run (v1.1.1)
-----------------------------------------------
[Mon May  4 11:07:23 2009] Preparing output location ./tophat_out/
[Mon May  4 11:07:23 2009] Checking for Bowtie index files
[Mon May  4 11:07:23 2009] Checking for reference FASTA file
[Mon May  4 11:07:23 2009] Checking for Bowtie
	Bowtie version:		 0.9.9.1
[Mon May  4 11:07:23 2009] Checking reads
	seed length:	 75bp
	format:		 fastq
	quality scale:	 phred
	Splitting reads into 3 segments
[Mon May  4 11:07:23 2009] Mapping reads against test_ref with Bowtie
[Mon May  4 11:07:24 2009] Mapping reads against test_ref with Bowtie
[Mon May  4 11:07:24 2009] Mapping reads against test_ref with Bowtie
	Splitting reads into 3 segments
[Mon May  4 11:07:24 2009] Mapping reads against test_ref with Bowtie
[Mon May  4 11:07:24 2009] Mapping reads against test_ref with Bowtie
[Mon May  4 11:07:24 2009] Mapping reads against test_ref with Bowtie
[Mon May  4 11:07:24 2009] Searching for junctions via coverage islands
[Mon May  4 11:07:24 2009] Searching for junctions via mate-pair closures
[Mon May  4 11:07:24 2009] Retrieving sequences for splices
[Mon May  4 11:07:24 2009] Indexing splices
[Mon May  4 11:07:24 2009] Mapping reads against segment_juncs with Bowtie
[Mon May  4 11:07:24 2009] Mapping reads against segment_juncs with Bowtie
[Mon May  4 11:07:24 2009] Mapping reads against segment_juncs with Bowtie
[Mon May  4 11:07:24 2009] Joining segment hits
[Mon May  4 11:07:24 2009] Mapping reads against segment_juncs with Bowtie
[Mon May  4 11:07:24 2009] Mapping reads against segment_juncs with Bowtie
[Mon May  4 11:07:24 2009] Mapping reads against segment_juncs with Bowtie
[Mon May  4 11:07:24 2009] Joining segment hits
[Mon May  4 11:07:24 2009] Reporting output tracks
-----------------------------------------------
Run complete [00:00:00 elapsed]
		</pre>
		<p>
		In the directory <tt>tophat_out</tt> should be a file <tt>junctions.bed</tt>.
		This file should contain a pair of junctions, on the reference sequence
		"test_chromosome".
	    </p>
		<br/>
		<h2 id="ref">Preparing your reference</h2><br/>
		
		<p> To find junctions with TopHat, you'll first need to <a
        href="http://bowtie-bio.sourceforge.net/tutorial.html#preb">install
        a Bowtie index</a> for the organism in your RNA-Seq experiment. The
        Bowtie site provides pre-built indices for human, mouse, fruit fly,
        and others. If there's no index for your organism, it's easy to <a
        href="http://bowtie-bio.sourceforge.net/tutorial.html#newi">build one
        yourself. </a> </p><br/>
		
		<p> TopHat also requires a fasta file
        (.fa) for your reference. If this file is not found alongside the 
		other index files, the program will use the Bowtie
		index you give it to build this file and save it to the output directory.  This 
		step can take up to an hour for a human-sized genome.  To skip this
		step in future runs, you can move the fasta file from the 
		<tt>tophat_out</tt> directory to the directory containing the Bowtie index
		files. 
		</p>

		<br/>
    	<h2 id="reads">Preparing your reads</h2><br/>
      	
		<p>TopHat currently accepts reads in FASTA or FASTQ format, though
        FASTQ is recommended. You may need to convert your reads from another
        format to one of these. Maq's <tt>fq_all2std.pl</tt> converts many
        formats into FASTQ. For reads in <a
        href="http://srf.sourceforge.net/">SRF format</a>, we recommend using
        the tools bundled with the <a
        href="https://sourceforge.net/project/showfiles.php?group_id=100316&package_id=108243&release_id=340318">
        Staden io_lib</a> package. </p>
        
        <p> <strong>Note:</strong> TopHat does not support mixing FASTA and
        FASTQ reads in the same input file, so <strong>don't</strong>
        run TopHat on FASTQ and FASTA files in the same run. </p>
        <br/>
      	
      	<h2 id="toph">Running TopHat</h2><br/>
      	<p> TopHat will map your reads first by running Bowtie to identify
        places where reads map end to end. Since your reads came from spliced
        transcripts in an RNA-Seq experiment, Bowtie will identify "islands"
        in your reference genomewhere reads piled up. Many of these islands
        will be exons. </p>
        
        <p> TopHat will then run a program to find splice junctions using the
        reads that did not get mapped to an island. So to identify junctions,
        you do <strong>not</strong> need to run Bowtie yourself, as TopHat
        will do it for you.</p>
                
        <p> TopHat needs you specify a path to the <a href="#ref">index
        files</a> and an input file containing your reads. The first argument
        should be the full path to the directory containing the index plus the
        <strong>prefix</strong> of the index files. To start the TopHat
        pipeline, enter the command: </p>
        
        <br/>
        <blockquote>tophat /path/to/h_sapiens reads1.fq,reads2.fq,reads3.fq</blockquote>
        <br/>
        
        <p> 
            Be sure to check out the TopHat <a href="./manual.html">manual</a>,
            as the pipeline has a few options you might want to use to get 
            better results or get them more quickly.
        </p> 
      	<br/>
      	
      	<h2 id="exam">Examining your output</h2><br/>
      	<p>TopHat produces <a href="./manual.html#output">several files </a> of 
	    output.  Because TopHat reports output in widely adopted formats, you can import it directly into a number of genome browsers and data viewers, including <a href="http://www.broadinstitute.org/igv/">IGV</a>, <a href="http://www.bioviz.org/igb/">IGB</a>, and the <a href="http://genome.ucsc.edu/">UCSC genome browser</a>.
	 	</p>
		<br/>
  </div>
  <div id="footer">
  	<table width="100%" cellspacing=15><tr><td>
    This research was supported in part by NIH grants R01-LM06845 and R01-GM083873 and NSF grant CCF-0347992.
    </td><td align=right>
    Administrator: <a href="mailto:cole@cs.umd.edu">Cole Trapnell</a>. Design by <a href="http://www.free-css-templates.com" title="Design by David Herreman">David Herreman</a>
    </td></tr></table>
  </div>
</div>

<!-- Google analytics code -->
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
var pageTracker = _gat._getTracker("UA-6101038-1");
pageTracker._trackPageview();
</script>

<!-- End google analytics code -->

</body>
</html>
